hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
  * License along with this library; if not, see <http://www.gnu.org/licenses/>.
  */
 
-#pragma once
+#ifndef BLE_HAL_H
+#define BLE_HAL_H
 
-#include "ble_hal_impl.h"
+#include "hal_platform.h"
 
-#include "system_error.h"
+#if HAL_PLATFORM_BLE
 
 #include <stdint.h>
 #include <stddef.h>
 
-// TODO: This API is not ready for general purpose usage
+#include "ble_hal_defines.h"
+#include "ble_hal_impl.h"
+
+/**
+ * @addtogroup ble_api
+ *
+ * @brief
+ *   This module provides BLE HAL APIs.
+ * @{
+ */
+
+typedef enum {
+    BLE_ANT_DEFAULT  = 0,
+    BLE_ANT_INTERNAL = 1,
+    BLE_ANT_EXTERNAL = 2,
+} hal_ble_ant_type_t;
+
+/* BLE device address */
+typedef struct {
+    uint8_t addr_type;
+    uint8_t addr[BLE_SIG_ADDR_LEN];
+} hal_ble_addr_t;
+
+/* BLE UUID, little endian */
+typedef struct {
+    uint8_t type;
+    union {
+        uint16_t uuid16;
+        uint8_t uuid128[BLE_SIG_UUID_128BIT_LEN];
+    };
+} hal_ble_uuid_t;
+
+/* BLE advertising parameters */
+typedef struct {
+    uint8_t version;
```

Comment:
This is a general comment about versioning: I personally prefer to also have `size`, because it simplifies and makes certain things safe out of the box, e.g.:
```cpp
struct api_struct {
    uint16_t version;
    uint16_t size;

    int a;
    int b;
    int c;

    // Version 2 fields
    int d;
};

void api_function_get_api_struct(api_struct* s) {
    // API implementation already uses version = 2, but the api_struct passed was version = 1, which is smaller
    memcpy(s, internal_v2_api_struct, s->size); // This is safe, will not write out of bounds, and no special handling is required (although would override size and version passed originally
}

void caller() {
    // Compiled with version = 1 API
    api_struct s = {};
    s.version = 1;
    s.size = sizeof(s); // sizeof before version 2 fields were added
    api_function_get_api_struct(&s);
}
```

Just something to keep in mind.